### PR TITLE
`leancode_hooks` improvements

### PR DESCRIPTION
**Pull request by bartekpacia**
- fix link in flutter_hooks readme
- fix leancode_hooks CHANGELOG links
- start adding tests for hooks (only for `useThrottle` for now)

### CODE REVIEW

**Comment by bartekpacia** [comment]
one more thing: gotta add GH Actions workflow for leancode_hooks test

File: packages/leancode_hooks/test/use_throttle_test.dart [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+      expect(find.text(textB), findsOneWidget);
+
+      await tester.pumpAndSettle(duration);
+      await Future<void>.delayed(duration);
+
+      await tester.tap(text);
+      await tester.pump();
+
+      expect(find.text(textC), findsOneWidget);
+    });
```
Comment (Albert221) [inline packages/leancode_hooks/test/use_throttle_test.dart:70]: Trailing comma 🙏

File: packages/leancode_hooks/test/use_throttle_test.dart [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+      home: GestureDetector(
+        onTap: () => throttle(updateText),
+        child: Text(state.value),
+      ),
+    );
+  }
+}
+
+void main() {
+  testWidgets('blocks calls to debounce while throttle is in progress',
```
[the remark this answers is not in the document]
Comment (bartekpacia) [reply packages/leancode_hooks/test/use_throttle_test.dart]: done

**Comment by Albert221** [comment]
Please use Flutter CLI for testing as this package depends on Flutter, not sole Dart

**Comment by bartekpacia** [comment]
I copy-pasted testes from `login_client` 🥲

File: packages/leancode_hooks/CHANGELOG.md [hunk 3]
```diff
@@ -1,11 +1,14 @@
...
+
+- Add tests for `useThrottle`
+- Fix broken links in README and CHANGELOG.
+
 # 0.0.2
 
 - Improve README.
 - Delete `useGoogleMapController` hook. You can view it
-  [here](google_map_controller).
+  [here](https://gist.github.com/bartekpacia/3cac94d29f8e6e438e57b366a742f73e).
```
Comment (shilangyu) [inline packages/leancode_hooks/CHANGELOG.md:10]: We should have a centralized place for all external hooks (readme?)

Comment (bartekpacia) [reply packages/leancode_hooks/CHANGELOG.md:10]: I don't think README is good for this. Some hooks might be long. 

But it's still betther than some gist, I admit.

Comment (shilangyu) [reply packages/leancode_hooks/CHANGELOG.md:10]: > Some hooks might be long.

I just meant links to them :D

Comment (bartekpacia) [reply packages/leancode_hooks/CHANGELOG.md:10]: Ah, okay :)

Comment (Albert221) [reply packages/leancode_hooks/CHANGELOG.md:10]: We have it in Git history, please refrain from using private GitHub account

File: packages/leancode_hooks/lib/src/use_decarative_text_editing_controller.dart [hunk 4]
```diff
@@ -1,6 +1,10 @@
 import 'package:flutter/material.dart';
 import 'package:flutter_hooks/flutter_hooks.dart';
 
+/// A wrapper around [useTextEditingController] that updates
+/// [TextEditingController.text] when it changes.
```
Comment (Albert221) [inline packages/leancode_hooks/lib/src/use_decarative_text_editing_controller.dart:5]: when _what_ changes?

File: packages/leancode_hooks/lib/src/use_decarative_text_editing_controller.dart [hunk 5]
```diff
@@ -1,6 +1,10 @@
...
+///
+/// Does not handle some edge cases with the current cursor/selection.
```
Comment (Albert221) [inline packages/leancode_hooks/lib/src/use_decarative_text_editing_controller.dart:7]: Could you explain a bit more about what cases are those and what happens then?

File: packages/leancode_hooks/lib/src/use_synced_text_editing_controller.dart [hunk 6]
```diff
@@ -1,6 +1,8 @@
 import 'package:flutter/widgets.dart';
 import 'package:flutter_hooks/flutter_hooks.dart';
 
+/// A wrapper around [useTextEditingController] that makes it easier to add
+/// [onChanged] callback to [TextEditingController].
 TextEditingController useSyncedTextEditingController(
```
Comment (Albert221) [inline packages/leancode_hooks/lib/src/use_synced_text_editing_controller.dart:6]: We now have 3 different hooks related to `TextEditingController` do we really need all of them?

File: packages/leancode_hooks/lib/src/use_decarative_text_editing_controller.dart (hunk 5 above)
Comment (bartekpacia) [reply packages/leancode_hooks/lib/src/use_decarative_text_editing_controller.dart:7]: @shilangyu 😬

**Comment by Albert221** [comment]
dev branch no longer exists in Flutter, we can remove testing with it

File: packages/leancode_hooks/lib/src/use_synced_text_editing_controller.dart (hunk 6 above)
Comment (bartekpacia) [reply packages/leancode_hooks/lib/src/use_synced_text_editing_controller.dart:6]: 3? I think 2: `useDeclatativeTextEditingController` and `useSyncedTextEditingController`

Comment (bartekpacia) [reply packages/leancode_hooks/lib/src/use_synced_text_editing_controller.dart:6]: and ideally these 2 would be merged into 1

Comment (Albert221) [reply packages/leancode_hooks/lib/src/use_synced_text_editing_controller.dart:6]: I also counted `useTextEditingController`

File: packages/leancode_hooks/lib/src/use_decarative_text_editing_controller.dart (hunk 5 above)
Comment (shilangyu) [reply packages/leancode_hooks/lib/src/use_decarative_text_editing_controller.dart:7]: TBH I would remove this hook, it was used for a very specific usecase in our project

**Comment by shilangyu** [comment]
> Apparently Flutter 2.17 made WidgetsBinding.instance non-nullable.

To make analyzer work on both beta and stable you need either a lint ignore or a wrapper function (which is what flutter packages often do). Example (from `package:camera`):

```dart
/// This allows a value of type T or T? to be treated as a value of type T?.
///
/// We use this so that APIs that have become non-nullable can still be used
/// with `!` and `?` on the stable branch.
// TODO(ianh): Remove this once we roll stable in late 2021.
T? _ambiguate<T>(T? value) => value;
```

File: packages/leancode_hooks/lib/src/use_decarative_text_editing_controller.dart (hunk 5 above)
Comment (bartekpacia) [reply packages/leancode_hooks/lib/src/use_decarative_text_editing_controller.dart:7]: I'd do this in `v0.0.4`

**Review by shilangyu — APPROVED** [review]

**Pull request merged**